### PR TITLE
Refactor: Runif for TPU and Horovod 5/n

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

add argument for TPU and Horvod accel and propagate it

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/helpers/runif.py [hunk 1]
```diff
@@ -83,6 +109,18 @@ def __new__(
...
+            reasons.append("TPU")
+
+        if horovod:
+            conditions.append(not _HOROVOD_AVAILABLE)
+            reasons.append("Horovod")
+
+        if horovod_nccl:
+            conditions.append(not _HOROVOD_NCCL_AVAILABLE)
+            reasons.append("Horovod with NCCL")
+
```
Comment (carmocca) [inline tests/helpers/runif.py:123]: The reasons above are `missing X`

and these are just `X`

choose one 🙂

File: tests/loggers/test_tensorboard.py [hunk 2]
```diff
@@ -25,12 +25,10 @@
...
 from pytorch_lightning.loggers import TensorBoardLogger
 from tests.helpers import BoringModel
+from tests.helpers.runif import RunIf
 
 
-@pytest.mark.skipif(
-    LooseVersion(torch.__version__) < LooseVersion("1.5.0"),
-    reason="Minimal PT version is set to 1.5",
-)
+@RunIf(min_torch="1.5")
```
Comment (carmocca) [inline tests/loggers/test_tensorboard.py:31]: You should use the full version, see why:

```
LooseVersion("1.5.0") == LooseVersion("1.5")
Out[13]: False
```

Comment (Borda) [reply tests/loggers/test_tensorboard.py:31]: good point
```python
LooseVersion("1.5.0") < LooseVersion("1.5")
Out[7]: False
LooseVersion("1.5.0") > LooseVersion("1.5")
Out[8]: True
```

File: tests/helpers/runif.py (hunk 1 above)
Comment (Borda) [reply tests/helpers/runif.py]: done

File: tests/helpers/runif.py [hunk 3]
```diff
@@ -51,6 +69,9 @@ def __new__(
             quantization: if `torch.quantization` package is required to run test
```
Comment (SkafteNicki) [inline tests/helpers/runif.py:69]:
```suggestion
            min_python: minimum python version required to run test
            quantization: if `torch.quantization` package is required to run test
```

File: tests/helpers/runif.py [hunk 4]
```diff
@@ -66,6 +87,11 @@ def __new__(
             conditions.append(torch_version < LooseVersion(min_torch))
             reasons.append(f"torch>={min_torch}")
 
+        if min_python:
+            py_version = f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
+            conditions.append(py_version < LooseVersion(min_python))
+            reasons.append(f"python>={min_python}")
```
Comment (SkafteNicki) [inline tests/helpers/runif.py:93]:
```suggestion
            reasons.append(f"requires python>={min_python}")
```

**Review by SkafteNicki — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: tests/helpers/runif.py (hunk 4 above)
Comment (Borda) [reply tests/helpers/runif.py:93]: requires is already in the group test `Requires: [...]`

File: tests/helpers/runif.py [hunk 5]
```diff
@@ -83,6 +109,18 @@ def __new__(
             conditions.append(sys.platform == "win32")
             reasons.append("unimplemented on Windows")
 
+        if tpu:
+            conditions.append(not _TPU_AVAILABLE)
+            reasons.append("missing TPU")
```
Comment (Borda) [inline tests/helpers/runif.py:114]:
```suggestion
            reasons.append("TPU")
```

File: tests/helpers/runif.py [hunk 6]
```diff
@@ -83,6 +109,18 @@ def __new__(
...
+
+        if horovod:
+            conditions.append(not _HOROVOD_AVAILABLE)
+            reasons.append("missing Horovod")
```
Comment (Borda) [inline tests/helpers/runif.py:118]:
```suggestion
            reasons.append("Horovod")
```

File: tests/helpers/runif.py [hunk 7]
```diff
@@ -83,6 +109,18 @@ def __new__(
...
+
+        if horovod_nccl:
+            conditions.append(not _HOROVOD_NCCL_AVAILABLE)
+            reasons.append("missing Horovod with NCCL")
```
Comment (Borda) [inline tests/helpers/runif.py:122]:
```suggestion
            reasons.append("Horovod with NCCL")
```

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_gpu_stats_monitor.py [hunk 8]
```diff
@@ -68,7 +67,7 @@ def test_gpu_stats_monitor(tmpdir):
         assert any([f in h for h in met_data.dtype.names])
 
 
-@pytest.mark.skipif(torch.cuda.is_available(), reason="test requires CPU machine")
+@RunIf(min_gpus=1)
```
Comment (Borda) [inline tests/callbacks/test_gpu_stats_monitor.py:70]:
```suggestion
@pytest.mark.skipif(torch.cuda.is_available(), reason="test requires CPU machine")
```

File: tests/models/test_amp.py [hunk 9]
```diff
@@ -162,7 +162,7 @@ def test_amp_gpu_ddp_slurm_managed(tmpdir):
     assert generated == 'abc23'
 
 
-@pytest.mark.skipif(torch.cuda.is_available(), reason="test is restricted only on CPU")
+@RunIf(min_gpus=1)
```
Comment (Borda) [inline tests/models/test_amp.py:165]:
```suggestion
@pytest.mark.skipif(torch.cuda.is_available(), reason="test is restricted only on CPU")
```

File: tests/accelerators/test_ddp.py [hunk 10]
```diff
@@ -82,8 +82,7 @@ def test_cli_to_pass(tmpdir, args=None):
     return '1'
 
 
-@RunIf(skip_windows=True)
-@pytest.mark.skipif(torch.cuda.is_available(), reason="test doesn't requires GPU machine")
+@RunIf(min_gpus=1, skip_windows=True)
```
Comment (Borda) [inline tests/accelerators/test_ddp.py:85]:
```suggestion
@RunIf(skip_windows=True)
@pytest.mark.skipif(torch.cuda.is_available(), reason="test doesn't requires GPU machine")
```

**Pull request merged**